### ISSUE TITLE
Ref'd 'anyOf' schemas only generates first 5 types

### ISSUE THREAD

**Issue by simonwebb-spafax** [bcherny/json-schema-to-typescript#273]
We are using a schema to include other schemas to generate, e.g.

Schema
```{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "$id": "schema/lookup.schema.json",
  "title": "Lookup Data",
  "anyOf": [
    {
      "$ref": "rating-authority.schema.json"
    },
    {
      "$ref": "region.schema.json"
    },
    {
      "$ref": "country.schema.json"
    },
    {
      "$ref": "distributor.schema.json"
    },
    {
      "$ref": "managing-lab.schema.json"
    },
    {
      "$ref": "content-tag.schema.json"
    },
    {
      "$ref": "rating-authorityrating.schema.json"
    }
  ]
}
```
However the output only includes the first 5 in the anyOf list, i.e.

```
export type LookupData = Region | Country | Distributor | ManagingLab | ContentTag;
[etc etc]
...
```
Is this a know issue and is there a workaround?

**Comment by JulianSobott** [comment]
I think the problem is when there are multiple types with the same attributes. In this case, only one of them is generated.

This schema is a good example:

```json
{
  "title": "Test",
  "type": "object",
  "oneOf": [
    {
      "type": "object",
      "title": "p2",
      "properties": {
        "name": {
          "type": "string"
        }
      }
    },
    {
      "type": "object",
      "title": "p1",
      "properties": {
        "name": {
          "type": "string"
        }
      }
    }
  ]
}
```

**Actual output:**

```ts
export type Test = P2;

export interface P2 {
  name?: string;
  [k: string]: unknown;
}
```

As you can see, `P1` is not generated. @simonwebb-spafax can you validate, that you have references to types, where they have the same attributes?

**Wanted/Expected output:**

```ts
export type Test = P2 | P1;

export interface P1 {
  name?: string;
  [k: string]: unknown;
}

export interface P2 {
  name?: string;
  [k: string]: unknown;
}
```

## Workaround

Any change to one of the types solves the problem.

- different property name
- one with: `"additionalProperties": false`
- one with required
- maybe more

**Comment by bcherny** [comment]
I took a quick pass at this.

@simonwebb-spafax @JulianSobott @joemitchard @mlradaji is this is the output you're looking for?

After #294:

[Input](https://github.com/bcherny/json-schema-to-typescript/pull/294/files#diff-e35feea50200e06700e833af3e27a608):

```ts
{
  title: 'Test',
  type: 'object',
  definitions: {
    a: {
      type: 'object',
      title: 'a',
      properties: {
        name: {
          type: 'string'
        }
      }
    },
    b: {
      type: 'object',
      title: 'b',
      properties: {
        name: {
          type: 'string'
        }
      }
    },
    c: {
      type: 'number',
      title: 'c'
    },
    d: {
      type: 'number'
    }
  },
  properties: {
    a: {
      oneOf: [{$ref: '#/definitions/a'}]
    },
    b: {
      oneOf: [{$ref: '#/definitions/a'}, {$ref: '#/definitions/b'}]
    },
    c: {
      oneOf: [{$ref: '#/definitions/c'}, {$ref: '#/definitions/d'}]
    },
    d: {
      oneOf: [{$ref: '#/definitions/c'}, {$ref: '#/definitions/c'}, {$ref: '#/definitions/d'}]
    },
    e: {
      oneOf: [{$ref: '#/definitions/d'}, {$ref: '#/definitions/d'}]
    },
    f: {
      oneOf: [
        {$ref: '#/definitions/a'},
        {$ref: '#/definitions/b'},
        {$ref: '#/definitions/c'},
        {$ref: '#/definitions/d'},
        {$ref: '#/definitions/a'},
        {$ref: '#/definitions/b'},
        {$ref: '#/definitions/c'},
        {$ref: '#/definitions/d'}
      ]
    }
  }
}

```

[Output](https://github.com/bcherny/json-schema-to-typescript/pull/294/files#diff-031e2da9d0269ecfb8d55d7742f4d00eR2156):

```ts
export type C = number;
export type D = number;

export interface Test {
  a?: A;
  b?: A | B;
  c?: C | D;
  d?: C | D;
  e?: D;
  f?: A | B | C | D;
  [k: string]: unknown;
}
export interface A {
  name?: string;
  [k: string]: unknown;
}
export interface B {
  name?: string;
  [k: string]: unknown;
}
```

**Comment by JulianSobott** [comment]
@bcherny this is exactly the output I was looking for. Thanks for the quick response.

**Comment by simonwebb-spafax** [comment]
@JulianSobott  - Thanks for taking a look. I've checked the references and some of the types do have a common 'Name' property but generally they are quite different so maybe it similar behaviour but not the same cause?

One of our devs managed to get it working by using a workaround of wrapping an 'allOf' around the 'anyOf', i.e.

```
  "$id": "schema/lookup.schema.json",
  "title": "Lookup Data",
  "allOf": [
    {
      "anyOf": [
        {
          "$ref": "region.schema.json"
        },
        {
          "$ref": "country.schema.json"
        },
        {
          "$ref": "distributor.schema.json"
        },
        {
          "$ref": "managing-lab.schema.json"
        },
        {...
```

**Comment by simonwebb-spafax** [comment]
@bcherny - Thanks for the response. I'll double check with our code as soon as I can

**Comment by bcherny** [comment]
Published 9.1.0.
